### PR TITLE
feat: create backend filter

### PR DESCRIPTION
**Pull request by Mikazian**
## Description

- Create a new migration for create a new table 'preference'
- Add new where clause in the SELECT users to matching with the user's preferences
- Create a PUT route to update user's preferences
- Add to GET route users the logic to display the list of user with matching logic
- Create middlewares to get the user's preference and filter

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Area of change

- [ ] Frontend
- [x] Backend
- [ ] Other: .....

## General checklist:

- [x] My code follows the style guidelines of this project
- [x] I ran `bun run format`/`bun run lint` before commit
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation where needed
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any misspellings
- [x] I have updated `develop` and merged to my branch before submitting pull request
- [x] My pull request generate no conflicts with `develop` branch
- [x] I requested code review from other team members

## Frontend checklist:

- [ ] I followed guidelines for `HTML`/`JSX`/`Typescript`
- [ ] My `Javascript` generate no new console errors
- [ ] I tested my code cross browsers
- [ ] My slice is pixel perfect for both desktop and mobile according to design
- [ ] I conducted basic QA to assure all features are working
- [ ] I tested responsive for mobile and tablet resolutions

## Backend checklist:

- [x] I tested admin by manually adding content from zero
- [x] I created easy to use admin experience which is self-explanatory
- [x] I added description to admin fields in hard-to-understand areas
- [x] I followed guidelines for naming `Typescript` variables
- [x] I conducted basic QA to assure all features are working

### CODE REVIEW

File: packages/shared/src/preference.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+export interface UserPreferenceId {
+  preference_id: number;
+}
+
+// Schema to validate the request body for updating preferences
+export const requestPreferencesSchema = z.object({
+  distance: z.number().int().nonnegative(),
+  gender_pref: z.enum(['man', 'woman', 'non-binary']),
+  language_pref_id: z.number().int().positive(),
```
Comment (greegouze) [inline packages/shared/src/preference.ts:11]: pourquoi distance nonnegative, et languages_pref_id positive ? positive pour distance ne fonctionne pas ?

**Review by greegouze — APPROVED** [review]

**Review by ElBarto444 — DISMISSED** [review]

**Review by greegouze — APPROVED** [review]

**Review by greegouze — APPROVED** [review]

**Review by greegouze — APPROVED** [review]

File: packages/backend/api/src/middlewares/filter-handlers.ts
Comment (YDrogen) [inline packages/backend/api/src/middlewares/filter-handlers.ts:1]: C'est pas un middleware ça, ça serait plus un service

File: packages/backend/api/src/middlewares/user-handlers.ts
Comment (YDrogen) [inline packages/backend/api/src/middlewares/user-handlers.ts:1]: idem pas un middleware

File: packages/backend/api/src/user/user-crud.ts [hunk 2]
```diff
@@ -40,14 +44,21 @@ userRouter.get('/:userId/profile', getUserId, async (req: Request, res) => {
...
 // Fetch a list of users without the user logged in
-userRouter.get('/:userId', getUserId, getUsers, (req: Request, res) => {
-  try {
-    const users = req.usersList;
+userRouter.get(
+  '/:userId',
+  getUserId,
+  getUserPreferences,
+  getUsers,
+  filteredUsersByDistance,
```
Comment (YDrogen) [inline packages/backend/api/src/user/user-crud.ts:52]: C'est pas censé être trop utilisé comme middleware
Conception un peu étrange

**Review by YDrogen — CHANGES_REQUESTED** [review]

File: packages/backend/api/package.json [hunk 3]
```diff
@@ -7,11 +7,12 @@
     "build": "bun build ./src/index.ts --target bun --minify --splitting --outdir ./dist"
   },
   "dependencies": {
-    "@app/shared": "workspace:*",
     "@app/backend-shared": "workspace:*",
+    "@app/shared": "workspace:*",
```
[the remark this answers is not in the document]
Comment (Mikazian) [reply packages/backend/api/package.json:11]: bun run lint && bun run format, j'ai rien touché ici

File: packages/backend/migrations/src/migrations/20240112083058-database-structure.ts
[the remark this answers is not in the document]
Comment (Mikazian) [reply packages/backend/migrations/src/migrations/20240112083058-database-structure.ts:1]: J'ai opté pour la 2e solution, les enlevé preference_id de la table user et mis user_id dans la table preference puis j'ai ajusté les données préférences du dossier forge

File: packages/backend/api/src/middlewares/filter-handlers.ts
Comment (Mikazian) [reply packages/backend/api/src/middlewares/filter-handlers.ts:1]: J'ai supprimé le middleware pour le transformer en service

File: packages/backend/api/src/middlewares/user-handlers.ts
Comment (Mikazian) [reply packages/backend/api/src/middlewares/user-handlers.ts:1]: J'ai supprimé le middleware pour le transformer en service

File: packages/backend/api/src/user/user-crud.ts (hunk 2 above)
Comment (Mikazian) [reply packages/backend/api/src/user/user-crud.ts:52]: C'est retiré, j'appelle les services à l'intérieur des routes

**Review by greegouze — APPROVED** [review]

**Review by YDrogen — APPROVED** [review]

**Pull request merged**